core/src/main/java/org/apache/accumulo/core/data/Value.java

```diff
@@ -242,16 +242,16 @@ public int compareTo(final byte[] that) {
 
   @Override
   public boolean equals(Object right_obj) {
-    // compare with byte[] for backwards compatibility, but this is generally a pretty bad practice
-    if (right_obj instanceof byte[]) {
-      return compareTo((byte[]) right_obj) == 0;
```

Comment:
Could throw an exception in this case.  That way if anyone was relying on this behavior it fails cleanly.

Reply:
I thought about that, but also thought that could easily break things in surprising ways. Also, the portion of the API contract for `equals` that I was trying to fix was specifically the "symmetry" portion. Throwing an exception here leaves that broken, undermining what I was trying to accomplish.

Also, I think people typically expect the question of equivalence to have an actual answer. Even null parameters are expected to have an answer and not throw NPE.

I also considered adding a log message, but... this object is a simple "value object" POJO, which probably shouldn't have a logger (logging for those kinds of objects should really occur by the caller where the object is used).

Reply:
Any code that was relying on this will break in surprising ways that are hard to debug or even detect.  The worst case for this change is that it causes a silent failure in user code that corrupts data that may not be detected for long periods of time.  The exception is a solution to this worst case.

Throwing an exception is not ideal, but neither was this original code.   The code has been like this for close to 10 years.  For that time period, its not unreasonable to think that a project may have relied on this behavior.

Reply:
I understand the problematic consequences of a change in behavior, but I don't see the point in throwing an exception, when that leaves the original problem in place (non-symmetrical equals implementations). It would be better to not do anything, I think.

My intent was to rely on the fact that Accumulo does not use this behavior, it was undocumented behavior, and *very* unlikely to be used by anybody else. Given those conditions, and with sufficient documentation in the release notes, I was thinking a breaking change in behavior in a 2.0.0 release would be fine. But, if the consensus is that it's not worth the risk, it's no big deal. We don't have to fix this at all.

Reply:
I am ok with making this change (w/o the exception), since the behavior was not documented.  Thinking about it some more there are potential negative consequences with making the change and not making the change.  I have no idea which will be most beneficial overall.

Reply:
oh, if  the change is made its nice the release notes can be edited now.

Reply:
Also since this class lacks javadoc, the current impl violates the javadoc from the parent class.